llvm/lib/Transforms/Vectorize/VPlan.h

```diff
@@ -3662,6 +3665,13 @@ class VPRegionBlock : public VPBlockBase {
   /// VPRegionBlock.
   VPBlockBase *Exiting;
 
+  /// Hold the Early Exit block of the SEME region, if one exists.
+  VPBasicBlock *EarlyExit;
+
+  /// If one exists, this keeps track of the vector early mask that triggered
+  /// the early exit.
+  VPValue *VectorEarlyExitCond;
+
   /// An indicator whether this region is to generate multiple replicated
```

Comment:
IIUC those are only needed because currently the final values are computed outside VPlan, right?

Could we instead either force a wide IV and extract the desired lane and add the required recipes directly on VPlan-construction?

Or compute the end value for the user directly in VPlan (see #112145 which computes the end values for the resume phis  #112147 which removes `fixupIVUsers`)?

Reply:
I've rewritten this so that we don't need to stash the `EarlyExit` and `VectorEarlyExitCond` variables. Instead, I've added a `getEarlyExit` that walks through the VPBasicBlocks to find the vector.early.exit block. I also added code in `addUsersInExitBlocks` to work backwards from the early exit block to find the early exit mask. It's uglier, but it does avoid having to track these variables and maintain them during cloning, etc. Hope that's prefereable for you!

Reply:
Thanks, I still need to take a look at the latest update. By forcing a wide IV I meant is ensuring that there always will be a vector induction, from which we should be able to extract the lane for the early exit I think.

One way to ensure that would be to not mark induction updates as uniform if there's an user of the induction outside the loop for early exit loops (something like https://github.com/llvm/llvm-project/commit/5c0d7aad84d52a6afcbd3feb4eeffde5ff94a89d)

Reply:
Just double checking, looking at the latest version it is effectively doing an extract of the last active lane, right?